### PR TITLE
qemu_guest_agent: support get_vcpus in windows guest

### PR DESCRIPTION
**Pull request by xiagao**
1.refactor gagent_check_get_vcpus function.
2.support windows guest.
ID: 1481508
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
(1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (44.21 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (58.52 s)

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
+        error_context.context("Check cpu info from guest agent.", logging.info)
+        vcpus_info = self.gagent.get_vcpus()
+        cpu_num_qga = len(vcpus_info)
+        for vcpu in vcpus_info:
+            if params.get("os_type") == "linux" and not vcpu["can-offline"]:
+                test.fail("Linux guest cpu can't be offline from qga"
+                          "which isn't expected.")
+            if self.params.get("os_type") == "windows" and vcpu["can-offline"]:
+                test.fail("Windows guest cpu can be offline from qga"
+                          "which isn't expected.")
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao If some vcpus cann't be offlined, some can be offlined.   As testing will break once hit cann't offline's vcpu. So,Maybe some can-offline vcpus will be skipped.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
         """
-        self.gagent.get_vcpus()
+        error_context.context("Check cpu info from guest agent.", logging.info)
+        vcpus_info = self.gagent.get_vcpus()
+        cpu_num_qga = len(vcpus_info)
+        for vcpu in vcpus_info:
+            if params.get("os_type") == "linux" and not vcpu["can-offline"]:
+                test.fail("Linux guest cpu can't be offline from qga"
+                          "which isn't expected.")
+            if self.params.get("os_type") == "windows" and vcpu["can-offline"]:
```
Comment (ybduan) [inline qemu/tests/qemu_guest_agent.py]: Please unify the way of getting os_type, Line 471 and Line 474

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: windows guest can't do offline operation,else fail this case.

File: qemu/tests/qemu_guest_agent.py (hunk 2 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: updated.

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 3]
```diff
@@ -84,8 +84,10 @@
                     new_password = "123456"
                     crypted = "yes"
         - check_get_vcpus:
-            no Windows
             gagent_check_type = get_vcpus
+            get_cpu_cmd = "ls /sys/devices/system/cpu |grep -i cpu[0-9]"
```
Comment (xutian) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: use  'lscpu' command instead please.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
+        Steps:
+        1) Check cpu info from guest agent.
+        2) Check cpu number.
+
         :param test: kvm test object
         :param params: Dictionary with the test parameters
         :param env: Dictionary with test environment.
         """
-        self.gagent.get_vcpus()
+        error_context.context("Check cpu info from guest agent.", logging.info)
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: it looks this step should be "Check offline cpus from guest agent "

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
+        if params.get("os_type") == "windows":
+            cpu_list = output.strip().split('\n')[1:]
+            cpu_num_guest = 0
+            for num in cpu_list:
+                cpu_num_guest += int(num)
+        else:
+            cpu_num_guest = len(output.strip().split('\n'))
+
+        if cpu_num_qga != cpu_num_guest:
+            test.fail("Cpu number is not correct.\n"
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: sed -i 's/Cpu/CPU/g'

Comment (xutian) [reply qemu/tests/qemu_guest_agent.py]: test.fail("CPU number mismatch,  number from guest-agent: %d, number from OS: %s"")

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
+                      " number in guest is %s,in qga is %s." %
+                      (cpu_num_guest, cpu_num_qga))
+        session.close
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: session.close(), right?

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @xiagao linux guest has the same issue.  In addition, This patch cann't cover porlarion case.  For example.  porlation case includes "offline vcpus" and "negative test".

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: @xiangchunfu  you mean linux guest, some vcpu can be offlined,some can be not? As I know, the filed "can-oofline" is fixed by qemu-ga, if not, I think it's qemu-ga product bug.
And the polarion case is covered in function gagent_check_set_vcpus().

Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: For function test. Looks good to me.
Acked by: xfu@redhat.com

**Comment by xiagao** [comment]
@xutian  Modified, could you help review? thanks.
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (32.02 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (41.48 s)

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 7]
```diff
@@ -84,8 +84,10 @@
                     new_password = "123456"
                     crypted = "yes"
         - check_get_vcpus:
-            no Windows
             gagent_check_type = get_vcpus
+            get_cpu_cmd = "lscpu |egrep -i '^CPU\(s\)' |awk -F: {'print $2'}"
```
Comment (xutian) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: lscpu |awk -F: '/^CPU\(s\):/ {print $2}'

Comment (xiagao) [reply qemu/tests/cfg/qemu_guest_agent.cfg]: update, thx.

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 8]
```diff
@@ -84,8 +84,10 @@
                     new_password = "123456"
                     crypted = "yes"
         - check_get_vcpus:
-            no Windows
             gagent_check_type = get_vcpus
+            get_cpu_cmd = "lscpu |awk -F: '/^CPU\(s\)/{print $2}'"
+            Windows:
+                get_cpu_cmd = "wmic cpu get NumberOfCores"
```
Comment (xutian) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: wmic cpu get NumberOfCores |findstr /r "[0-9]"

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -454,12 +454,44 @@ def gagent_check_set_user_password(self, test, params, env):
...
+                          "which isn't expected.")
+
+        error_context.context("Check cpu number.", logging.info)
+        session = self._get_session(params, self.vm)
+        output = session.cmd_output(params["get_cpu_cmd"])
+
+        if params.get("os_type") == "windows":
+            cpu_list = output.strip().split('\n')[1:]
+            cpu_num_guest = 0
+            for num in cpu_list:
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: cpu_num_guest = sum(map(int, cpu_list))

**Comment by xiagao** [comment]
> Ok, let's make code clean.

Thanks. Modified.
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (34.21 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (378.73 s)

File: qemu/tests/qemu_guest_agent.py [hunk 10]
```diff
@@ -454,12 +454,42 @@ def gagent_check_set_user_password(self, test, params, env):
...
+                test.fail("Linux guest cpu can't be offline from qga"
+                          "which isn't expected.")
+            if params.get("os_type") == "windows" and vcpu["can-offline"]:
+                test.fail("Windows guest cpu can be offline from qga"
+                          "which isn't expected.")
+
+        error_context.context("Check cpu number.", logging.info)
+        session = self._get_session(params, self.vm)
+        output = session.cmd_output(params["get_cpu_cmd"])
+
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: move session.close() to here

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: > move session.close() to here

update, thanks.

File: qemu/tests/qemu_guest_agent.py [hunk 11]
```diff
@@ -454,12 +454,42 @@ def gagent_check_set_user_password(self, test, params, env):
     @error_context.context_aware
     def gagent_check_get_vcpus(self, test, params, env):
         """
-        Execute "guest-set-vcpus" command to guest agent
+        Execute "guest-get-vcpus" command to guest agent.
+
+        Steps:
+        1) Check can-offline field of guest agen.
```
Comment (hereischen) [inline qemu/tests/qemu_guest_agent.py]: agent？

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: > agent？

update.thx

**Comment by xiagao** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (30.94 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_get_vcpus: PASS (381.27 s)

**Pull request merged**